### PR TITLE
Add 'allow' and 'disallow' setting to each manager

### PR DESCRIPTION
**Pull request by shreekarSS**
Signed-off-by: Shreekara <sshreeka@redhat.com>

## Related Issues and Dependencies
Fixes: https://github.com/thoth-station/kebechet/issues/1126
…

## This Pull Request implements
New config setting for managers in .thoth.yaml
For ex: 

managers:
  - name: thoth-advise
	//Do not run this manager in the ps-nlp-tensorflow-gpu overlay
	env_disallow_list:
	- ps-nlp-tensorflow-gpu
	configuration:
	  labels: [bot]
  - name: update
       //Run this manager in the ps-nlp-tensorflow-gpu overlay (only)
	env_allow_list:
	- ps-nlp-tensorflow-gpu
	configuration:
	  labels: [bot]
  - name: info
	//No allow/disallow present: this manager applies to all runtimes
...

### CODE REVIEW

File: kebechet/managers/thoth_advise/thoth_advise.py [hunk 1]
```diff
@@ -316,6 +316,15 @@ def run(self, labels: list, analysis_id=None):
                     runtime_environments = [
                         e["name"] for e in thoth_config["runtime_environments"]
                     ]
+                    managers = thoth_config.get("managers")
+                    disallow_run_times = managers[0].get("disallow", None)
+                    allowed_run_times = managers[0].get("allow", None)
+                    if disallow_run_times:
+                        runtime_environments = list(
+                            set(runtime_environments) ^ set(disallow_run_times)
```
Comment (KPostOffice) [inline kebechet/managers/thoth_advise/thoth_advise.py:324]: I think this should just be set subtraction `-`

File: kebechet/managers/thoth_advise/thoth_advise.py [hunk 2]
```diff
@@ -316,6 +316,15 @@ def run(self, labels: list, analysis_id=None):
...
+                        )
+                    if allowed_run_times:
+                        runtime_environments = allowed_run_times
```
Comment (KPostOffice) [inline kebechet/managers/thoth_advise/thoth_advise.py:327]: I think this would be better as set intersection `&`

File: kebechet/managers/thoth_advise/thoth_advise.py [hunk 3]
```diff
@@ -316,6 +316,15 @@ def run(self, labels: list, analysis_id=None):
                     runtime_environments = [
                         e["name"] for e in thoth_config["runtime_environments"]
                     ]
+                    managers = thoth_config.get("managers")
+                    disallow_run_times = managers[0].get("disallow", None)
```
Comment (KPostOffice) [inline kebechet/managers/thoth_advise/thoth_advise.py:320]: This should be per manager, not global to all managers

**Review by KPostOffice — COMMENTED** [review]
Good start!

I think it might make sense to make managerbase take runtime_environments as a list and have it calculated this way.

See [here](https://github.com/thoth-station/kebechet/blob/356af8ecf138bb22b0642044d20e86a3d405c11a/kebechet/kebechet_runners.py#L215)

File: kebechet/managers/thoth_advise/thoth_advise.py (hunk 3 above)
Comment (shreekarSS) [reply kebechet/managers/thoth_advise/thoth_advise.py:320]: ack, thanks

File: kebechet/managers/thoth_advise/thoth_advise.py (hunk 2 above)
Comment (shreekarSS) [reply kebechet/managers/thoth_advise/thoth_advise.py:329]: ack, done

File: kebechet/managers/thoth_advise/thoth_advise.py [hunk 4]
```diff
@@ -316,6 +316,21 @@ def run(self, labels: list, analysis_id=None):
                     runtime_environments = [
                         e["name"] for e in thoth_config["runtime_environments"]
                     ]
+                    managers = thoth_config.get("managers")
+                    for mgr in managers:
+                        if mgr.get("name") == "thoth-advise":
+                            disallow_run_times = mgr.get("disallow", None)
+                            allowed_run_times = mgr.get("allow", None)
```
Comment (KPostOffice) [inline kebechet/managers/thoth_advise/thoth_advise.py:323]: This shouldn't have to be done here, should be passed through the managers configuration variables.

File: kebechet/kebechet_runners.py [hunk 5]
```diff
@@ -172,11 +172,26 @@ def run(
         return
 
     managers = config.managers
-
+    if config.config.get("overlays_dir"):
+        runtime_environment = [
```
Comment (KPostOffice) [inline kebechet/kebechet_runners.py:176]: Let's change the name to runtime_environments

File: kebechet/kebechet_runners.py [hunk 6]
```diff
@@ -172,11 +172,26 @@ def run(
...
+            e["name"] for e in config.config.get("runtime_environments")
+        ]
+    else:
+        runtime_environment = [config.config["runtime_environments"][0]["name"]]
     for manager in managers:
         # We do pops on dict, which changes it. Let's create a soft duplicate so if a user uses
         # YAML references, we do not break.
         manager = dict(manager)
+        disallow_run_times = manager.get("disallow", None)
+        allowed_run_times = manager.get("allow", None)
```
Comment (KPostOffice) [inline kebechet/kebechet_runners.py:186]: These are technically mutually exclusive. Not sure how important it is to users though.

File: kebechet/kebechet_runners.py [hunk 7]
```diff
@@ -172,11 +172,26 @@ def run(
         return
```
Comment (KPostOffice) [inline kebechet/kebechet_runners.py:173]: We need to respect the passed runtime_environment. We could also allow the user to pass a list of runtime_environments in the future, I don't think we need to do this now though.

**Review by KPostOffice — COMMENTED** [review]
Looks good. A few comments

File: kebechet/managers/thoth_advise/thoth_advise.py [hunk 8]
```diff
@@ -311,51 +311,43 @@ def run(self, labels: list, analysis_id=None):
...
-                            )
-                        )
-                    except (FileLoadError, ThothConfigurationError) as exc:
-                        if isinstance(exc, FileLoadError):
-                            self._tracking_issue.comment(
-                                f"""Result for {e}: Error advising, no requirements found.
-
-                                If this project does not use requirements.txt or Pipfile then remove thoth-advise
-                                manager from your .thoth.yaml configuration."""
+                if self.runtime_environments:
```
Comment (KPostOffice) [inline kebechet/managers/thoth_advise/thoth_advise.py:314]: This check shouldn't be necessary if the default would be an empty list. If the default is `None` however, you could change the for loop to:
```
for e in self.runtime_environments or []:
```

Here and in update manager

**Review by KPostOffice — COMMENTED** [review]
One comment in two spots, otherwise it looks good!

File: kebechet/kebechet_runners.py (hunk 7 above)
Comment (shreekarSS) [reply kebechet/kebechet_runners.py:173]: ack

File: kebechet/managers/thoth_advise/thoth_advise.py (hunk 8 above)
Comment (shreekarSS) [reply kebechet/managers/thoth_advise/thoth_advise.py:314]: ack done

**Review by harshad16 — CHANGES_REQUESTED** [review]
Please update documentation for this feature in this PR itself.

**Comment by shreekarSS** [comment]
> Please update documentation for this feature in this PR itself.

Ack, thanks ,fixed

File: docs/managers.rst [hunk 9]
```diff
@@ -85,3 +85,63 @@ Kebechet works as a part of Thoth Ecosystem, please raise an issue or add the
 new manager to the `KebechetGithubAppInstallations
 <https://github.com/thoth-station/storages/blob/15ed39ef6c8d7bf58037046f3bab2465c5c4bb22/thoth/storages/graph/models.py#L1434>`_
 table.
+
+
+Configure different managers per Overlays
+-----------------------------------------
+
+- Add allow and disallow to each manager's configuration options, where the user can specify which repos each manager is (dis)allowed to act on.
```
Comment (harshad16) [inline docs/managers.rst:93]:
```suggestion
- Add allow and disallow to each manager's configuration options, where the user can specify which manager is (dis)allowed for each overlay.
```

File: kebechet/kebechet_runners.py [hunk 10]
```diff
@@ -172,11 +172,27 @@ def run(
...
+            e["name"] for e in config.config.get("runtime_environments")
+        ]
+    else:
+        runtime_environments = [config.config["runtime_environments"][0]["name"]]
     for manager in managers:
         # We do pops on dict, which changes it. Let's create a soft duplicate so if a user uses
         # YAML references, we do not break.
         manager = dict(manager)
+        disallow_run_times = manager.get("disallow", None)
+        allowed_run_times = manager.get("allow", None)
```
Comment (harshad16) [inline kebechet/kebechet_runners.py:187]: for default: empty list `[]` might be better than `None`.

File: kebechet/kebechet_runners.py [hunk 11]
```diff
@@ -172,11 +172,27 @@ def run(
...
+        if disallow_run_times:
+            runtime_environments = list(
+                set(runtime_environments) - set(disallow_run_times)
+            )
+        if allowed_run_times:
+            runtime_environments = list(
+                set(runtime_environments).intersection(set(allowed_run_times))
+            )
```
Comment (harshad16) [inline kebechet/kebechet_runners.py:195]: we should check on some earlier points to validate the schema and if the allow_run_time and disalow_run_times match the existing runtimes, if not open the issue for the user.

now for this code block:
will a simple if work here:
```
if allowed_run_times:
runtime_environments = allowed_run_time 
```

File: kebechet/managers/thoth_advise/thoth_advise.py [hunk 12]
```diff
@@ -311,51 +311,43 @@ def run(self, labels: list, analysis_id=None):
...
                     thoth_config = yaml.safe_load(f)
-
-                if thoth_config.get("overlays_dir"):
-                    runtime_environments = [
-                        e["name"] for e in thoth_config["runtime_environments"]
-                    ]
-                else:
-                    runtime_environments = [
-                        thoth_config["runtime_environments"][0]["name"]
-                    ]
```
Comment (harshad16) [inline kebechet/managers/thoth_advise/thoth_advise.py:322]: why is part not needed anymore?

are we checking somewhere else, if the repo is overlays or normal type?

**Review by harshad16 — COMMENTED** [review]
thanks for the great work. 
dropped a few comments.

File: kebechet/managers/thoth_advise/thoth_advise.py (hunk 12 above)
Comment (shreekarSS) [reply kebechet/managers/thoth_advise/thoth_advise.py:322]: Yes, this code block is moved to https://github.com/thoth-station/kebechet/blob/c54451adaabe27190a749354ce0acb7a414cf339/kebechet/kebechet_runners.py#L176 to remove code redundancy

File: docs/managers.rst (hunk 9 above)
Comment (shreekarSS) [reply docs/managers.rst]: Fixed, thanks

File: kebechet/managers/thoth_advise/thoth_advise.py (hunk 12 above)
Comment (harshad16) [reply kebechet/managers/thoth_advise/thoth_advise.py:322]: sounds good. thanks

**Review by harshad16 — COMMENTED** [review]
Thanks 
/lgtm

PTAL @KPostOffice when you get some time.

File: kebechet/managers/update/update.py [hunk 13]
```diff
@@ -715,19 +715,7 @@ def run(self, labels: list = []) -> Optional[dict]:
 
             runtime_environments: List[Optional[str]]
             if self.runtime_environment:
-                if self.runtime_environment not in runtime_environment_names:
-                    # This is not a warning as it is expected when users remove and change runtime_environments
-                    _LOGGER.info("Requested runtime does not exist in target repo.")
-                    return None
-                runtime_environments = [self.runtime_environment]
```
[the remark this answers is not in the document]
Comment (shreekarSS) [reply kebechet/managers/update/update.py]: fixed

File: kebechet/kebechet_runners.py (hunk 10 above)
Comment (shreekarSS) [reply kebechet/kebechet_runners.py]: ack, done

File: kebechet/kebechet_runners.py [hunk 14]
```diff
@@ -172,11 +172,29 @@ def run(
         return
 
     managers = config.managers
-
+    runtime_environments: List[str]
+    if runtime_environment:
+        runtime_environments = [runtime_environment]
+    if config.config.get("overlays_dir"):
```
Comment (KPostOffice) [inline kebechet/kebechet_runners.py:178]: elif here

File: docs/managers.rst [hunk 15]
```diff
@@ -85,3 +85,62 @@ Kebechet works as a part of Thoth Ecosystem, please raise an issue or add the
...
+          - name: ps-nlp-tensorflow-gpu
+            operating_system:
+              name: ubi
+            version: "8"
+            python_version: "3.8"
+            recommendation_type: latest
+        managers:
+          - name: thoth-advise
+            # Do not run this manager in the ps-nlp-tensorflow-gpu overlay
+            disallow:
```
Comment (KPostOffice) [inline docs/managers.rst:135]: I'm thinking we should make the name more clear. What do you think about `env_allow_list` and `env_disallow_list`?

File: kebechet/managers/update/update.py [hunk 16]
```diff
@@ -109,6 +109,7 @@ def __init__(self, *args, **kwargs):
         # We do API calls once for merge requests and we cache them for later use.
         self._cached_merge_requests = None
         self._pr_list = []
+        self.runtime_environment = "default"
```
Comment (KPostOffice) [inline kebechet/managers/update/update.py:112]: Why is this line added here?

File: kebechet/managers/update/update.py [hunk 17]
```diff
@@ -710,26 +711,14 @@ def run(self, labels: list = []) -> Optional[dict]:
...
                     _LOGGER.info("Requested runtime does not exist in target repo.")
-                    return None
-                runtime_environments = [self.runtime_environment]
-            else:
-                if overlays_dir:
-                    runtime_environments = runtime_environment_names
-                elif runtime_environment_names:
-                    runtime_environments = [runtime_environment_names[0]]
-                else:
-                    runtime_environments = [None]
```
Comment (KPostOffice) [inline kebechet/managers/update/update.py:729]: I think we lost the "default" functionality here.  I think the for loop should probably be default `[None]` instead of `[]`

**Review by KPostOffice — COMMENTED** [review]
A few more comments

File: kebechet/managers/update/update.py (hunk 16 above)
Comment (shreekarSS) [reply kebechet/managers/update/update.py:112]: previously , self.runtime_environment was a string, being used in https://github.com/thoth-station/kebechet/blob/master/kebechet/managers/update/update.py#L761 and many more places.
Since we are converting it list of strings, do not want to break the code since it's already being used in various methods, created a variable to keep it running

File: kebechet/managers/thoth_advise/thoth_advise.py (hunk 8 above)
Comment (codificat) [reply kebechet/managers/thoth_advise/thoth_advise.py:314]: I still see the `if` statement in the current PR. I think it can be handled in the `for` as @KPostOffice mention.

File: kebechet/managers/thoth_advise/thoth_advise.py [hunk 18]
```diff
@@ -311,51 +311,43 @@ def run(self, labels: list, analysis_id=None):
 
                 with open(".thoth.yaml", "r") as f:
                     thoth_config = yaml.safe_load(f)
-
-                if thoth_config.get("overlays_dir"):
-                    runtime_environments = [
```
Comment (codificat) [inline kebechet/managers/thoth_advise/thoth_advise.py:316]: As we are removing the local `runtime_environments` var, let's also remove its declaration above (line 297)

**Review by codificat — CHANGES_REQUESTED** [review]
Found one minor thing remaining

File: kebechet/managers/thoth_advise/thoth_advise.py (hunk 18 above)
Comment (shreekarSS) [reply kebechet/managers/thoth_advise/thoth_advise.py:316]: ack,done

**Review by harshad16 — APPROVED** [review]
/lgtm

lets get this in :+1:
